libs/wasm-detective/src/keywords/scanner.rs

```diff
@@ -0,0 +1,297 @@
+                    f.value = value.to_string();
+                }
+            }
+
+            f.pii_matches.extend(self.match_against_fields(&f));
+
+            let mut confidence_sum: f32 = 0.0;
+            f.pii_matches.iter().for_each(|m| {
+                confidence_sum += m.confidence;
+            });
+            if confidence_sum > 100.0 {
+                confidence_sum = 100.0;
+            }
+            f.confidence = confidence_sum;
+
+            f.pii_matches.iter_mut().for_each(|m| {
+                m.confidence = confidence_sum;
+            });
+
+            fields.push(f);
+
+            true
+        });
+
+        fields
+    }
+
+    /// Match a field against our keyword lists
+    /// There are two lists computed at initialization:
+    /// 1. scalar_keywords: A list of single keywords to match against
+    /// 2. path_keywords: A list of keywords with periods in them to match against. These represent
+    ///          a path in the JSON payload
+    /// scalar_keywords will be a O(1) lookup
+    /// strict path_keywords will be a O(1) lookup
+    /// fuzzy path_keywords will be a O(n) lookup
+    fn match_against_fields(&mut self, f: &Field) -> Vec<KeywordMatch> {
+        let mut matches: Vec<KeywordMatch> = Vec::new();
+
+        // Single keyword lookup
+        //
+        // Example payload {"config": {"github": {"token": "pat_90210"}}}
+        // Example key match: "token"
+        //
+        // O1 lookup
+        let key_name = clean_key(f.key_name.to_lowercase().as_str());
+        if self.scalar_keywords.contains_key(&key_name) {
+            let def = self.scalar_keywords.get(&key_name).unwrap(); // safe unwrap
+
+            let m = KeywordMatch {
+                path: f.path(),
+                value: f.value.clone(),
+                confidence: def.score as f32,
+                entity: def.entity.clone(),
+            };
+
+            matches.push(m);
+            return matches;
+        }
+
+        // Path keys get cleaned by path
+        let cur_path = f.path();
+
+        // Explicit path check here
+        // We didn't match on a single key such as "cvv" or "ssn". So let's check
+        // our list of explicit paths. These are keywords with a period in them
+        //
+        // Example payload {"config": {"github": {"token": "pat_90210"}}}
+        // Example key match: "config.github.token"
+        //
+        // O(1) lookup
+        if self
+            .path_keywords
+            .contains_key(cur_path.trim_start_matches('.'))
+        {
+            let def = self.path_keywords.get(cur_path.as_str()).unwrap(); // safe unwrap
+
+            let m = KeywordMatch {
+                path: f.path(),
+                value: f.value.clone(),
+                confidence: def.score as f32,
+                entity: def.entity.clone(),
+            };
+
+            matches.push(m);
+            return matches;
+        }
+
+        // Partial match check here
```

Comment:
Hmm.. maybe this should be configurable in the step config - ie. something like `EnablePartialPathMatching` - that way we don't take the hit by the default.

Reply:
Let's architect the protos for this. We'll also need fields in the future for custom keyword lists, so best to do it properly now